presto-orc/src/main/java/com/facebook/presto/orc/StreamLayout.java

```diff
@@ -0,0 +1,101 @@
+
+import com.facebook.presto.orc.metadata.Stream;
+import com.facebook.presto.orc.stream.StreamDataOutput;
+
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+import java.util.stream.Collectors;
+
+import static java.util.Objects.requireNonNull;
+
+/**
+ * StreamLayout is used by the Writer to reorder the data streams.
+ * In ORC, a column is stored in multiple streams (for example
+ * DATA, PRESENT, LENGTH, DICTIONARY and so on).
+ */
+public interface StreamLayout
+{
+    void reOrder(List<StreamDataOutput> dataStreams);
+
+    /**
+     * Streams are ordered by the Stream Size. There is no ordering between
+     * two streams of the same size. This strategy places all small streams
+     * together, so that multiple of these small IOs can be combined. Note:
+     * usually columns contain small (e.g. PRESENT) and large streams (e.g. DATA).
+     * This strategy places them far apart and may result in increased IO.
+     */
+    class ByStreamSize
+            implements StreamLayout
+    {
+        @Override
+        public void reOrder(List<StreamDataOutput> dataStreams)
+        {
+            requireNonNull(dataStreams, "dataStreams is null");
+            Collections.sort(dataStreams);
+        }
+    }
+
+    /**
+     * Streams are ordered by Column Size. If two columns have same size then
+     * columnId, stream size and stream kind are used for ordering. This strategy
+     * places small columns together to read multiple small columns in one IO.
+     * All streams for a column are stored together to read one column in one IO.
+     */
+    class ByColumnSize
+            implements StreamLayout
+    {
+        @Override
+        public void reOrder(List<StreamDataOutput> dataStreams)
+        {
+            requireNonNull(dataStreams, "dataStreams is null");
+            if (dataStreams.isEmpty()) {
+                return;
+            }
+
+            final Map<Integer, Long> columnSizes = dataStreams.stream()
+                    .collect(Collectors.toMap(
+                            s -> s.getStream().getColumn(),
+                            s -> (long) s.getStream().getLength(),
+                            (s1, s2) -> s1 + s2));
+
+            dataStreams.sort((sd1, sd2) -> {
+                Stream s1 = sd1.getStream();
+                Stream s2 = sd2.getStream();
+
+                long sizeDelta = columnSizes.get(s1.getColumn()) -
+                        columnSizes.get(s2.getColumn());
+                if (sizeDelta != 0) {
+                    return sizeDelta < 0 ? -1 : 1;
+                }
+
+                // order by column attributes
+                int columnDelta = s1.getColumn() - s2.getColumn();
+                if (columnDelta != 0) {
+                    return columnDelta;
+                }
+
+                sizeDelta = s1.getLength() - s2.getLength();
+                if (sizeDelta != 0) {
+                    return sizeDelta < 0 ? -1 : 1;
+                }
+
```

Comment:
Do we need to take care of Stream.sequence too?

Reply:
Presto Writer has no support for flat map, so sequence is unused for now. It would be weird to add sequence checks, when the Sequence is unused, so did not add the check.